library/eddsa.c

```diff
@@ -0,0 +1,446 @@
+}
+#endif
+
+/*
+ * Compute EdDSA signature of a message.
+ * For PREHASH operation, the message is already previously hashed.
+ * Obviously, for PREHASH, we skip hash message step.
+ */
+int mbedtls_eddsa_sign( mbedtls_ecp_group *grp,
+                mbedtls_mpi *r, mbedtls_mpi *s,
+                const mbedtls_mpi *d, const unsigned char *buf, size_t blen,
+                int (*f_rng)(void *, unsigned char *, size_t), void *p_rng,
+                mbedtls_eddsa_id eddsa_id,
+                const unsigned char *ed_ctx, size_t ed_ctx_len )
+{
+    int ret;
+    mbedtls_ecp_point Q, R;
+    mbedtls_mpi q, prefix, rq, h;
+
+    /* EdDSA only should be used with Ed25519 curve  */
+    if( ! mbedtls_eddsa_can_do( grp->id ) || grp->N.p == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+    if( grp->id == MBEDTLS_ECP_DP_ED25519 && eddsa_id != MBEDTLS_EDDSA_PURE && eddsa_id != MBEDTLS_EDDSA_CTX && eddsa_id != MBEDTLS_EDDSA_PREHASH )
+        return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+#endif
+
+    if( eddsa_id == MBEDTLS_EDDSA_PREHASH && blen != 64)
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    mbedtls_ecp_point_init( &Q ); mbedtls_ecp_point_init( &R );
+
+    mbedtls_mpi_init( &q ); mbedtls_mpi_init( &prefix ); mbedtls_mpi_init( &rq ); mbedtls_mpi_init( &h);
+
+    MBEDTLS_MPI_CHK( mbedtls_ecp_expand_edwards( grp, d, &q, &prefix ) );
+
+    MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &Q, &q, &grp->G, f_rng, p_rng ) );
+
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+    if( grp->id == MBEDTLS_ECP_DP_ED25519 )
+    {
+        mbedtls_sha512_context sha_ctx;
+        unsigned char sha_buf[64], tmp_buf[32];
+        size_t olen = 0;
+
+        /* r computation */
+        mbedtls_sha512_init( &sha_ctx );
+        mbedtls_sha512_starts( &sha_ctx , 0 );
+
+        if( eddsa_id == MBEDTLS_EDDSA_CTX )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 0, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+        else if( eddsa_id == MBEDTLS_EDDSA_PREHASH )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 1, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+
+        /* Update SHA with prefix */
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( &prefix, tmp_buf, sizeof(tmp_buf) ) );
+
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        /* In EDDSA_PREHASH, buf should contain the SHA512 hash. It contains the whole message otherwise */
+        mbedtls_sha512_update( &sha_ctx, buf, blen );
+
+        mbedtls_sha512_finish( &sha_ctx, sha_buf );
+        mbedtls_sha512_free( &sha_ctx );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( &rq, sha_buf, sizeof( sha_buf ) ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &rq, &rq, &grp->N ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &R, &rq, &grp->G, f_rng, p_rng ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_copy( r, &R.Y ) );
+        if( mbedtls_mpi_get_bit( &R.X, 0 ) )
```

Comment:
This is basically `mbedtls_ecp_point_write_binary()`, but done on MPI instead. Note that this function actually has a few additional safety check. Therefore I wonder if it could somehow be used. A possibility would be to pass `R` instead of `r` to `mbedtls_eddsa_sign()`

Reply:
Actually `mbedtls_ecp_point_write_binary()` encodes a point onto a buffer, but here we need to encode the point onto an mpi. Are these checks necessary if the point comes from the sign operation? In short, only the check `mbedtls_mpi_cmp_int( &P->Z, 1 ) != 0` could be necessary, but it always holds.

Reply:
I have seen that you added `mbedtls_ecp_point_encode()` and `mbedtls_ecp_point_decode()`. That sounds the good approach to me, and avoids having to use a buffer to do the conversion.